src/sdk/util.hpp

```diff
@@ -0,0 +1,155 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2020 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef realm_sdk_util_h
+#define realm_sdk_util_h
+
+#include <cxxabi.h>
+#include <type_traits>
+#include "property.hpp"
+
+namespace realm::sdk::util {
+
+namespace type_traits {
+struct embedded_object;
+struct object;
+struct list;
+struct primitive;
+struct enum_type;
+struct primary_key;
+} // namespace type_traits
+
+template <typename T>
+using EmbeddedObject = Property<T, util::type_traits::embedded_object>;
+
+static inline std::string demangle(const char* name) {
+    int status = -4; // some arbitrary value to eliminate the compiler warning
+
+    std::unique_ptr<char, void(*)(void*)> res {
+        abi::__cxa_demangle(name, NULL, NULL, &status),
+        std::free
+    };
+
+    return (status==0) ? res.get() : name;
+}
+
+template <typename, typename = void>
+struct is_embedded : std::false_type {};
+template <typename T>
+struct is_embedded<T, typename std::enable_if_t<std::is_base_of_v<EmbeddedObject<T>, T>>> : std::true_type {};
+template <typename T>
+static constexpr auto is_embedded_v = is_embedded<T>::value;
+
+template <class T, std::size_t = sizeof(T)>
+std::true_type is_complete_impl(T *);
+
+std::false_type is_complete_impl(...);
+
+template <class T>
+using is_complete = decltype(is_complete_impl(std::declval<T*>()));
+
+// trim from start (in place)
+static inline void ltrim(std::string &s) {
+    s.erase(s.begin(), std::find_if(s.begin(), s.end(), [](unsigned char ch) {
+        return !std::isspace(ch);
+    }));
+}
+
+// trim from end (in place)
+static inline void rtrim(std::string &s) {
+    s.erase(std::find_if(s.rbegin(), s.rend(), [](unsigned char ch) {
```

Comment:
Does this actually work? `std::string::reverse_iterator` should not be convertible to `std::string::iterator`.